rfcs/0000-rfc-forms.md

```diff
@@ -0,0 +1,148 @@
+---
+Status: Active
+Champions: christian24
+PR: {{ update_with_pr_number }}
+---
+
+# @lit-labs/forms
+
+Add advanced form validation for web components. This is a followup to the discussion in the lit repository: https://github.com/lit/lit/discussions/2489 
+
+## Objective
+
+Provide an easy to use dev experience for validating forms with web components in them. 
+
+### Goals
+- Provide a common API, so developers can quickly start using the @lit-labs/forms package without changing their existing controls. 
```

Comment:
I think the very first line and the main objective should be different: it's not only about validating forms, the main purpose of something like this would be having a single source of truth for a control (or group of controls) state (and being able to change it imperatively). And this state should include its value, its UI state (touched, dirty, blurred...) and its current validity (valid, invalid, pending).

To respond to @justinfagnani , under a different design people might have to use dedicated custom elements instead of native ones. Example: https://lion-web.netlify.app/fundamentals/systems/form/overview/

Preface: Lion's Form system is maybe the best solution for form management with WCs, I haven't used it yet but it seems very complete and it's probably the reason why I didn't open this RFC myself. But if I'm not mistaken, it's a template-driven mechanism and you have to use wrappers such as `<lion-input>` instead of just `<input>`.

A model-driven approach would keep the logic on the JS side and just bind the controls to an existing template.

Another thing that stopped me from opening an RFC was the fact that when I wrote the prototype, I realized how little of Lit was there in the source code, meaning that maybe it'd be better to have a vanillaJS solution with just a couple of adapters for Lit (Reactive Controllers, Directives) or any other framework. I stumbled upon Final Form (https://final-form.org/docs/final-form/philosophy), but I haven't evaluated it yet!

Reply:
This is a fine line we walk, but I think it's very ok to have a Lit-associated libraries that technically is mostly agnostic to Lit or ReactiveElement itself.

First, having a directive or reactive controller is reason enough. I think spending the time to figure out what an equivalent non-Lit patterns would be can come later. We did create the abstractions to be used...

Second, even if the library was very agnostic, it's ok to have it be under the Lit umbrella. Many developers don't quite grok the breakdown between general WC utilities and Lit-related things. We've gotten feedback that what a lot of people want is a single place to come for their component-related needs, much like a framework. In this sense Lit is a project and brand for a set of web component authoring helpers. You can already see this with some utilities that are technically pretty independent of Lit like the `@lit-labs/react`, `@lit-labs/task`, `@lit-labs/observers`...

### Model-driven

One thing I'd like to understand in this space is what role the model specifically plays... my concept of components is that they usually have data and events in; and data, DOM, and events out, and _hopefully_ minimal internal state. I would think that the data for a form is part of the data-in. I can see the need for local state for tracking transient form state, but I wonder:

1. How does that correlate to the data flowing into a component? Where do the property values merge with the form model? It's usually nice to keep state sync to other objects be to DOM children in `render()` because it's declarative. Would we sync properties to form model in `willUpdate()`?

2. Then what about persisting the transient state, say for navigation. If a user of an MPA (or SPA with internal navigation that disposes of the form) navigates away and back to the form, and the developer wants to restore their form state including "touched", etc., how would that work? Can the form model be put into a state management system like Redux that might be integrated with the router?

### Lion and Final Form

I think it would be great to have comparisons and alternatives in here. I haven't looked at either yet, but will do. @dakmor do you have any perspective on Lion forms?

Reply:
> To respond to @justinfagnani , under a different design people might have to use dedicated custom elements instead of native ones. Example: https://lion-web.netlify.app/fundamentals/systems/form/overview/
> 
> Preface: Lion's Form system is maybe the best solution for form management with WCs, I haven't used it yet but it seems very complete and it's probably the reason why I didn't open this RFC myself. But if I'm not mistaken, it's a template-driven mechanism and you have to use wrappers such as `<lion-input>` instead of just `<input>`.

Lion's form system inspired the form system we initially built, but your controls need to follow a specific API if I am not mistaken. 

> 
> A model-driven approach would keep the logic on the JS side and just bind the controls to an existing template.
> 

This is exactly what I find charming about a model-driven approach. 

>  I stumbled upon Final Form (https://final-form.org/docs/final-form/philosophy), but I haven't evaluated it yet!

I didn't know about this. Agree, we should compare it.

Reply:
> ### Model-driven
> One thing I'd like to understand in this space is what role the model specifically plays... my concept of components is that they usually have data and events in; and data, DOM, and events out, and _hopefully_ minimal internal state. I would think that the data for a form is part of the data-in. I can see the need for local state for 
Let me give you the example of using a simple `input` element. The data-in would be the `value` property. The data-out are the `change` or `input` events and the `value` property as well (depends on the implementation of your adapter). Every component will provide some API for this. However, in a real application, you will likely want to know if a control has been focused by the user, if it has had input, what its initial value is, etc. You cannot rely on having all of this info in the component (plus would be repetitive as well). That's mostly where the idea of the model comes from.
> 
> 1. How does that correlate to the data flowing into a component? Where do the property values merge with the form model? It's usually nice to keep state sync to other objects be to DOM children in `render()` because it's declarative. Would we sync properties to form model in `willUpdate()`?

I would argue that a part of the model is the data flowing into the component. But the model will also contain information that the component doesn't need to know about itself - touched, dirty or even the initial value of the component, because the component should only care about its current value. Updating the model will sync the component in the DOM and user input will sync to the model. So they should always be in sync. Not sure about doing this in `willUpdate()`, in our implementation we are doing it inside the directive. The model really is just data and you can subscribe to changes. It doesn't know of the component.

> 2. Then what about persisting the transient state, say for navigation. If a user of an MPA (or SPA with internal navigation that disposes of the form) navigates away and back to the form, and the developer wants to restore their form state including "touched", etc., how would that work? Can the form model be put into a state management system like Redux that might be integrated with the router?

All data in the model is basically a plain object. We are using `BehaviorSubject` to allow subscribing to changes. You can always access the the plain objects and also change them. So, yes, they can be stored in a store or even in a JSON store or wherever. The only place where this is a bit more complicated is form arrays (so dynamic forms), because from the array of values you cannot infer what kind of form group or form control needs to be created along side the data. This can be solved though by supplementing the constructor for form array with a function of how to create a new child (which can then be populated with the data). 
> 
> ### Lion and Final Form
> I think it would be great to have comparisons and alternatives in here. I haven't looked at either yet, but will do. @daKmoR do you have any perspective on Lion forms?

Agreed. 

@justinfagnani more of an out of interest question: Does the problem of how to handle forms never come up at Google? It was the first and probably biggest major road block we stumbled upon, so really surprised this doesn't come up more often.

Reply:
Hmmm I'm not sure I understand your first question @justinfagnani, why are we talking about properties? I mean, in a Lit app a form lives inside a component (or can be split in multiple components) but the form itself doesn't know about it, it's just internal state.

Are you talking about something like this?

```html
<my-component .foo=${'hey'}>
```

```ts
class MyComponent extends LitElement {

  @property() foo = '';

  render() {
    return html`<input .value=${this.foo}>`
  }
}
```

Because to me this is a very particular scenario (props to be synced with internal state) which almost never happens (I'd consider this a bad practice unless the value is only used for initialization). In cases like this we could use lifecycle methods (willUpdate / updated) to imperatively set a new value for the form.

---

For the second question, yes, it can be serialized and stored. In Angular I typically store long forms in localStorage / indexedDB (or Redux), an equivalent for Lit would be something like:

```ts
class ComponentWithForm extends LitElement {

  form = new FormGroup(...);

  connectedCallback() {
    // restore previous value
    const value = JSON.parse(localStorage.get('form'));
    this.form.setValue(value);

    // listen to the form
    this.stopListening = this.form.onValueChange(x => localStorage.set('form', JSON.stringify(x)));
  }

  disconnectedCallback() {
    this.stopListening?.();
  }
}
```

The `bind` directives takes care of updating the DOM with the restored values automatically, you never touch the template.

I never thought about restoring the UI state (touched, dirty...) but I guess it can be exposed by the Reactive Controller as a complex object to be stored, like:

```ts
{
  name: {
    type: 'control',
    value: 'hey',
    touched: true,
    dirty: true,
    ...
  }, ...
}
```

But this would be advanced stuff, I honestly never needed something like this :P

Reply:
We needed restoring UI state for multi step wizards. Every step was a page / route in the applications, so that when you press back in the browser you would just go to the previous page and your input would not be lost.

Reply:
Maybe additionally, the USP of a lit solution would be that although it has few dependencies on Lit itself, it would make sure that it works easily with lit. What comes to mind in that regard is that whenever we update component's inputs, we would need to call `requestUpdate()`. I am sure this can also be done with `final-form`, but with a lit solution this would be the "magic" that just works ;)